server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java

```diff
@@ -220,15 +233,52 @@ public CompletionStage<Collection<CacheDistributionInfo>> cacheDistribution(Stri
          return CompletableFuture.completedStage(Collections.singletonList(local));
       }
 
-      Map<Address, CacheDistributionInfo> distributions = new ConcurrentHashMap<>();
+      return requestMembers(dm.getCacheTopology().getMembers(), ecm -> {
+         Cache<?, ?> c = SecurityActions.getCache(cacheName, ecm);
+         return CacheDistributionInfo.resolve(c.getAdvancedCache());
+      });
+   }
+
+   public CompletionStage<CompleteKeyDistribution> getKeyDistribution(String cacheName, Object key, RestRequest request) {
+      AdvancedCache<Object, ?> cache = getCache(cacheName, request.keyContentType(), request);
+      checkCachePermission(cache, request.getSubject(), AuthorizationPermission.MONITOR);
+      return SecurityActions.cacheContainsKeyAsync(cache, key)
+            .thenCompose(exists -> {
+               if (!exists)
+                  return CompletableFuture.completedFuture(Collections.emptyList());
+
+               DistributionManager dm = SecurityActions.getDistributionManager(cache);
+               if (dm == null) {
+                  KeyDistributionInfo local = KeyDistributionInfo.resolve(cache, true);
+                  return CompletableFuture.completedFuture(Collections.singletonList(local));
+               }
+
+               LocalizedCacheTopology topology = dm.getCacheTopology();
+               DistributionInfo distribution = topology.getDistribution(key);
+               return requestMembers(distribution.readOwners(), ecm -> {
```

Comment:
Why are we going remote here? All of the information should be retrievable from the originating node.

Reply:
This is because we need the node name. I've created https://issues.redhat.com/browse/ISPN-14076 to improve this in the future.